From 58cdd03c9b3db4dca6ce79a32c184013a6231d3d Mon Sep 17 00:00:00 2001
From: Richard Zhu <r65037@freescale.com>
Date: Wed, 8 Oct 2008 10:56:08 +0800
Subject: [PATCH] ENGR00093235 Only PIO mode4 can be configured in 2.6.26 libata

There is a mode filter api in the 2.6.26, add another filter api
func to fix this issue.

Signed-off-by: Richard Zhu <r65037@freescale.com>
---
 drivers/ata/pata_fsl.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/drivers/ata/pata_fsl.c b/drivers/ata/pata_fsl.c
index ca2fe0c..987b591 100644
--- a/drivers/ata/pata_fsl.c
+++ b/drivers/ata/pata_fsl.c
@@ -467,6 +467,15 @@ static int pata_fsl_check_atapi_dma(struct ata_queued_cmd *qc)
 	return 1;		/* ATAPI DMA not yet supported */
 }
 
+unsigned long pata_fsl_bmdma_mode_filter(struct ata_device *adev,
+					unsigned long xfer_mask)
+{
+	/* Capability of the controller has been specified in the
+	 * platform data. Do not filter any modes, just return
+	 * the xfer_mask */
+	return xfer_mask;
+}
+
 static void pata_fsl_bmdma_setup(struct ata_queued_cmd *qc)
 {
 	int chan, i;
@@ -697,6 +706,7 @@ static struct ata_port_operations pata_fsl_port_ops = {
 
 	.check_atapi_dma = pata_fsl_check_atapi_dma,
 	.cable_detect = ata_cable_unknown,
+	.mode_filter = pata_fsl_bmdma_mode_filter,
 
 	.bmdma_setup = pata_fsl_bmdma_setup,
 	.bmdma_start = pata_fsl_bmdma_start,
-- 
1.5.4.4

